src/directive.ts

```diff
@@ -0,0 +1,279 @@
+/**
+ * @license
+ * Copyright (c) 2021 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive as legacyDirective} from './lib/directive.js';
+import * as legacyLit from './lit-html.js';
+
+export const PartType = {
+  ATTRIBUTE: 1,
+  CHILD: 2,
+  PROPERTY: 3,
+  BOOLEAN_ATTRIBUTE: 4,
+  EVENT: 5,
+  ELEMENT: 6,
+} as const ;
+
+export type PartType = typeof PartType[keyof typeof PartType];
+
+export interface ChildPartInfo {
+  readonly type: typeof PartType.CHILD;
+}
+
+export interface AttributePartInfo {
+  // eslint-disable-next-line @typescript-eslint/type-annotation-spacing
+  readonly type:|typeof PartType.ATTRIBUTE|
+      typeof PartType.PROPERTY|typeof PartType.BOOLEAN_ATTRIBUTE|
+      typeof PartType.EVENT;
+  readonly strings?: ReadonlyArray<string>;
+  readonly name: string;
+  readonly tagName: string;
+}
+
+export type Part = ChildPart|AttributePart|BooleanAttributePart|EventPart;
+
+type Interface<T> = {
+  [P in keyof T]: T[P];
+};
+
+export type ChildPart = Interface<ChildPartImpl>;
+class ChildPartImpl {
+  readonly type = PartType.CHILD;
+  readonly options: legacyLit.RenderOptions|undefined;
+  constructor(readonly legacyPart: legacyLit.NodePart) {
+    this.options = legacyPart.options;
+  }
+
+  get parentNode(): Node {
+    return this.legacyPart.startNode.parentNode!;
+  }
+}
+
+/**
+ * Yields all toplevel nodes inside the dom range managed by the ChildPart.
+ *
+ * So if the ChildPart has rendered:
+ *     html`<label><input></label> <div></div>`
+ *
+ * This function will yield the label and the div, but not the input.
+ */
+export function getRenderedNodes(childPart: ChildPart) {
+  const results = [];
+  const impl = childPart as ChildPartImpl;
+  const part = impl.legacyPart;
+  for (let node: Node|null = part.startNode; node && node !== part.endNode;
+       node = node.nextSibling) {
+    const n: Node = node;
+    results.push(n);
+  }
+  return results;
+}
+
+export type AttributePart = Interface<AttributePartImpl>;
+class AttributePartImpl {
+  readonly type: typeof PartType.ATTRIBUTE|typeof PartType.PROPERTY;
+
+  get options(): legacyLit.RenderOptions|undefined {
+    return undefined;
```

Comment:
For a very modest code size increase, we could go and put `options` on each legacy part type (the `DefaultTemplateProcessor` that creates all the parts has it) via an optional constructor argument on each Part class, so that the v2 parts can actually return something here.

It still won't guarantee it's there for people who created custom template processors, but seems like it would be an improvement (and they can't migrate that code to v2 anyway)?

Or maybe we just don't care because they couldn't have relied on options in v1, so it's not a migration concern?

@justinfagnani thoughts?

Reply:
IMO since this is for allowing existing directives to migrate to the new API surface, we don't need to support novel features, just translation

Reply:
Yeah, makes sense.

Reply:
I agree w/ @rictic here